datafusion/core/src/datasource/listing_table_factory.rs

```diff
@@ -49,16 +49,18 @@ impl ListingTableFactory {
 impl TableProviderFactory for ListingTableFactory {
     async fn create(
         &self,
-        state: &SessionState,
+        state: &dyn CatalogSession,
         cmd: &CreateExternalTable,
     ) -> Result<Arc<dyn TableProvider>> {
-        let file_format = state
+        // TODO remove downcast_ref from here. Should file format factory be an extension to session state?
```

Comment:
Not sure what do you mean by this 😖 
I think the role of FileFormat to SessionState is similar to function to SessionState, they are already registrable

Reply:
Good point. OTOH, `dyn TableProviderFactory` is already provided to the `ListingSchemaProvider`, so maybe the table provider could just have it, without getting it from the session?
Also, how do pluggable file format work with table formats such as Iceberg and Delta? Delta is always Parquet, but it's different Parquet metadata than Iceberg, so they are not handled the same way, are they?

Reply:
FWIW, `IcebergTableProvider` from https://github.com/apache/iceberg-rust
https://github.com/apache/iceberg-rust/blob/bbd042da5835b3e2e8ce7c05d45377c367464efd/crates/integrations/datafusion/src/table.rs#L76-L88 doesn't use the session at all.

i would be tempted to think that the pluggable FileFormat is equivalent to implementing Hive table format, where file format is underspecified and pluggable. Hive table format is super important industry-wide and maybe it could deserve it's own module (crate). Then,  the pluggable file format functionality could be its implementation detail and not necessarily part of DF core.

@jayzhan211 @alamb  please correct me if i am thinking about this in a wrong way

Reply:
also filed https://github.com/apache/datafusion/issues/11600 for this

Reply:
Moving datasource in it's own crate is our plan.

hive could be one of it too

- datafusion-datasource-parquet
- datafusion-datasource-csv
- datafusion-datasource-json
- datafusion-datasource-hive